### PR TITLE
pgconvert: option to specify output file

### LINKED ISSUE #11: FR pgconvert: optional second filename to specify output file
... ideally also for pgcollect

### PR DESCRIPTION
**Pull request by GitMensch**
fixes #11

first take (note: as my experience in c++ is limited, the code is likely not optimal - but it does work for:

* specifying the "special filename '-' -> (stdout)"
* specifying no filename ->(stdout)"
* specifying files that either don't exist and the user can write to the folder or files that are writable

### CODE REVIEW

**Comment by GitMensch** [comment]
I've applied the changes (and changed one comparison to follow the style in the rest of the file).
But I can't trigger the memory check with "out" that you've suggested to be automatically done:

```
perfgrind$ chmod 0 check_ls.grind
perfgrind$ ./pgconvert check_ls.pgdata -d object       1> check_ls.grind   # check by the OS
-bash: check_ls.grind: Permission denied
perfgrind$ ./pgconvert check_ls.pgdata -d object   check_ls.grind   # (no) check by stream -> goes to nirvana
Can't resolve symbol for address 15e31d, load base: 7f7a89ef8000
Can't resolve symbol for address 15e323, load base: 7f7a89ef8000
Can't resolve symbol for address 16133e, load base: 7f7a89ef8000
Can't resolve symbol for address 1617c5, load base: 7f7a89ef8000
```

File: pgconvert.cpp [hunk 1]
```diff
@@ -272,7 +283,13 @@ int main(int argc, char** argv)
 
   profile.resolveAndFixup(params.details);
 
-  dump(std::cout, profile, params.dumpInstructions);
-
+  if (strcmp("-", params.outputFile)) {
+    // misses check if file is writable
+    std::ofstream *out = new std::ofstream(params.outputFile);
+    dump(*out, profile, params.dumpInstructions);
+    out->close();
```
[the remark this answers is not in the document]
Comment (ostash) [reply pgconvert.cpp:289]: There is no need in dynamic allocation via `new`, but what what is really needed is to check whether `std::ofstream` constructed successfully via `if (out)`. Please, take a look on my suggested code.

**Review by ostash — CHANGES_REQUESTED** [review]

File: pgconvert.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ostash) [reply pgconvert.cpp:289]: There is also no need to close ofstream explicitly, as destructor will do that anyway.

[the remark this answers is not in the document]
Comment (GitMensch) [reply pgconvert.cpp:289]: Somehow that did not get commited, so doing again.

**Comment by ostash** [comment]
Please, use the commit message like this:

pgconvert: allow to specify output file

Next parameter after input file now specifies the output file. stdout will be used if not provided.

I want to start following rules like described here: https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

However, I do not want to create reference to Github (like closes #NN), as this make connection to GitHub too tight. What if I decide to move to Gitlab tomorrow? :)

**Comment by GitMensch** [comment]
I've adjusted the commit merge. In my personal opinion references to issues help to point to discussions like ours, they give additional context. If you move to GitLab, then the old issues would still be found here (but actually could be migrated, too).

I've taken the freedom to create the output buffer directly after the input file is parsed, so we won't get the output of profile loading and execute the dump only once, seems to work fine.

**Comment by GitMensch** [comment]
Note to your eee41888: the double-space at line endings in markdown like at https://github.com/ostash/perfgrind/compare/0996fbc99ff45a4966c2051ae66f5c43b741673a..eee418852d62b6a7c840104a9e39a8f153138073#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L48-R48 is actually important as otherwise rendering of that file won't have a line break there.

**Comment by ostash** [comment]
> Note to your eee41888: the double-space at line endings in markdown like at https://github.com/ostash/perfgrind/compare/0996fbc99ff45a4966c2051ae66f5c43b741673a..eee418852d62b6a7c840104a9e39a8f153138073#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L48-R48 is actually important as otherwise rendering of that file won't have a line break there.

Right, sorry. This one of the thing I do not like in markdown. Feel free to restore that space.

**Comment by GitMensch** [comment]
Done, so I think that's ready for merge, no?

**Comment by ostash** [comment]
> Done, so I think that's ready for merge, no?

No, dangling reference still needs to be fixed. As you mentioned you're not a C++ person, I can finish this for you if you want.

**Comment by GitMensch** [comment]
> No, dangling reference still needs to be fixed.

Would you mind to elaborate?

> As you mentioned you're not a C++ person, I can finish this for you if you want.

As you like, but as we've gone all that way together and I'm always eager to learn something new I'd prefer you mentoring this PR until I can provide "what is needed".

**Comment by ostash** [comment]
Ok, np. See, in C++ regular object (not dynamically-allocated) is destructed when its scope ends.

```c++
int* ptr = 0;
{
  int num = 5;
  // Saving address of num in ptr
  ptr = &num;
 // num is destroyed here 
}
// there is no num here, so ptr points to noones'/someone's else address
ptr = 10; // Depending on luck this is SIGSEGV (lucky case), writing to mapped memory not used by other objects (lucky case, which can become unlucky, or writing to some other object (unlucky case).
```

You can check that your code doesn't work :) Empty 0-sized output file is created, but not converted data is written there.

**Comment by GitMensch** [comment]
I see, the pointer does not count.
But then a `params.getOutputStream()` wouldn't work either, because the object created in there would leave the scope, no?

The _hack_ would be a conditional on the outer level and a go to after the opening (so the object does not leave the scope), a _workaround_ would be to let all the code that follows be duplicated, nearly as bad as the go to.

 So what is the solution?

**Comment by ostash** [comment]
> I see, the pointer does not count. But then a `params.getOutputStream()` wouldn't work either, because the object created in there would leave the scope, no?

Well, this could work somehow if `Params` were designed differently. Too much effort for such a small task.

> The _hack_ would be a conditional on the outer level and a go to after the opening (so the object does not leave the scope), a _workaround_ would be to let all the code that follows be duplicated, nearly as bad as the go to.
> 
> So what is the solution?

Either to redesign `Params` or switch to C++11 and use std::unqiue_ptr:
```c++
std::unique_ptr<std::ofstream> outPtr;
if (strcmp("-", params.outputFile))
{
  outPtr = new std::ofstream(params.outputFile);
  if (!*outPtr)
  { /* print error and exit */ }
}

dump(outPtr ? *outPtr : std::cout, params.dumpInstruction);
```

IMHO, just leave it as is, with dumping profile after loading and resolving it.

**Comment by GitMensch** [comment]
> IMHO, just leave it as is, with dumping profile after loading and resolving it.

That kind of hurts... but done as suggested.

**Comment by GitMensch** [comment]
Another thing done. So... is it likely that you'll find some time for #4 or is that unlikely? In the later case: Would you mind to tag 0.4 or is there anything else you'd like to include for that?

**Comment by ostash** [comment]
> Another thing done. So... is it likely that you'll find some time for #4 or is that unlikely? In the later case: Would you mind to tag 0.4 or is there anything else you'd like to include for that?

I will try to work on this in next week or two, but do not consider this as a hard promise :) I'm working on this during my free time.

Is lack of '-z' blocking you with test or usage?

I feel that just  '-o' for pgconvert is not enough for the release.

**Comment by GitMensch** [comment]
> Is lack of '-z' blocking you with test or usage?

quite possibly with the later, but not with the former

> I feel that just '-o' for pgconvert is not enough for the release

I see (note: there's already + software events + PLT size + fix compiler warnings + docs) for 0.4 :-)
But one or two weeks for additional -z won't harm.

**Comment by ostash** [comment]
Right, software events :) Everything else was cosmetics. We'll see how it will go.

**Pull request merged**